src/kernel/sporadic.c

```diff
@@ -248,94 +252,60 @@ static bool_t refill_head_overlapping(sched_context_t *sc)
+        userError(
+            "SC (%p) overran budget by %lluus (%llu ticks)",
+            sc,
+            ticksToUs(overrun),
+            overrun
+        );
     }
-
-    capacity = refill_capacity(sc, usage);
-    if (capacity > 0 && refill_ready(sc)) {
-        refill_split_check(usage);
+#endif
+    usage = MIN(usage, refill_head(sc)->rAmount);
+
+    /* Charge the usage to the head refill */
+    if (usage > 0) {
+        refill_t used = (refill_t) {
+            .rAmount = usage,
+            .rTime = refill_head(sc)->rTime + sc->scPeriod,
+        };
+
+        refill_head(sc)->rAmount -= usage;
+        /* We need to keep the head refill no more than a period before
+         * the start of the tail refill. This ensures that new refills
+         * are never added before the tail refill (breaking the ordered
+         * invariant). This code actually keeps the head refill no more
+         * than a period before the end of the tail refill (which is
+         * stronger than necessary) but is what is used in the current
+         * proofs. In combination with the merging behaviour of
+         * schedule_used, the following will still ensure that
+         * incremental charging of a refill across preemptions only
+         * produces a single new refill one period in the future. */
+        refill_head(sc)->rTime += usage;
+        schedule_used(sc, used);
     }
 
-    ensure_sufficient_head(sc);
-
-    REFILL_SANITY_END(sc);
-}
-
-void refill_split_check(ticks_t usage)
-{
-    sched_context_t *sc = NODE_STATE(ksCurSC);
-    /* invalid to call this on a NULL sc */
-    assert(sc != NULL);
-    /* something is seriously wrong if this is called and no
-     * time has been used */
-    assert(usage > 0);
-    assert(usage <= refill_head(sc)->rAmount);
-    assert(sc->scPeriod > 0);
-
-    REFILL_SANITY_START(sc);
+    /* Ensure the head refill has the minimum budget */
+    while (refill_head(sc)->rAmount < MIN_BUDGET) {
+        refill_t head = refill_pop_head(sc);
+        refill_head(sc)->rAmount += head.rAmount;
+        /* Delay head to ensure the subsequent refill doesn't end any
+         * later (rather than simply combining refills). */
+        refill_head(sc)->rTime -= head.rAmount;
+    }
 
-    /* first deal with the remaining budget of the current replenishment */
-    ticks_t remnant = refill_head(sc)->rAmount - usage;
-
-    /* set up a new replenishment structure */
-    refill_t new = (refill_t) {
-        .rAmount = usage, .rTime = refill_head(sc)->rTime + sc->scPeriod
-    };
-
-    if (refill_size(sc) == sc->scRefillMax || remnant < MIN_BUDGET) {
-        /* merge remnant with next replenishment - either it's too small
-         * or we're out of space */
-        if (refill_single(sc)) {
-            /* update inplace */
-            new.rAmount += remnant;
-            *refill_head(sc) = new;
-        } else {
-            refill_pop_head(sc);
-            refill_head(sc)->rAmount += remnant;
```

Comment:
I think this branch is necessary to handle the case when the refill queue is full (https://github.com/seL4/seL4/blob/master/include/kernel/thread.h#L258).  The part that is being removed would correctly result in refill_ready being false upon return from refill_budget_check inside chargeBudget.  This way when chargeBudget calls endTimeslice (https://github.com/seL4/seL4/blob/master/src/kernel/thread.c#L598) the thread doesn't get added to the release queue when it should.  (It takes the roundrobin path if there's no timeout fault handler).  Then it will keep looping through these steps upon any kernel entry until it uses up its available time I think.

Reply:
I don't fully understand everything here yet, so please bear with me...

Are you saying that when a non-round-robin thread enters the kernel with `refill_full` true, it should *always* end up `postpone`d? I can see that this is the case in what's currently on `master`, due to the branch you pointed out here. However, I think this is only the case if we assume that next refill does not overlap with the current one, and AFAICT, the code currently on `master` does *not* ensure non-overlapping refills. That's because there are a number of places where code on `master` adds to `rAmount` without adjusting `rTime` or checking whether the bigger refill overlaps with the next one.

If I understand what you're saying, the problem with the code in this PR is that non-round-robin threads that enter the kernel with enough time left to still be sufficient after usage is charged, but with `refill_full` true, will follow the round robin path. That is, the used portion will be taken from the front of the head refill and added to the end of the tail refill. And since the head remnant is still sufficient, the thread will go straight back into the ready queue.

Is that what you're saying? I can't tell whether this contradicts the last few words of the comment in `checkBudget` ("the rest of the budget is forfeit").

Another thing I noticed about this PR's version of `refill_budget_check`: When a thread enters the kernel with `refill_full` true, but where the head refill is *not* sufficient after the usage is charged, then after `refill_budget_check` returns, `refill_full` is no longer true for the current SC. That's because `schedule_used` is called first, which adds the used amount to the tail refill, and then the first `while` loop combines the remnant (which is still present as the head refill) with subsequent refills until it meets `MIN_BUDGET`. Is it intended that this situation always leads to the SC being *not* `refill_full`?

Reply:
Thinking some more about that `refill_full` condition in `checkBudget`... Is it really the case that any interrupt should bump the current thread if it's SC is `refill_full`?

Reply:
@mbrcknl I think so.

The logic is as follows:

We don't know at the start of the kernel if we are going to have to split the budget or not. Assume we do have to split it (meaning we're switching scheduling contexts). Now if there is not enough space for the split to happen, we have to consume the whole head refill. Now we don't have enough time left in the SC because we had to consume the whole head refill (could't split it, no space) and it might need to be put in the release queue, but we were supposed to do that at the start of the kernel. 

There might be a way to wrangle this in a way that doesn't make a thread with plenty of time in its head refill lose it, involving more precarious charging, but I think you either lose what's at the head (hence bumping the thread) or push back the next replenishment. Either way you lose.

Reply:
Right, I think I see. The code in this PR appears to take a third approach, which is to push back the tail refill, so the used portion goes to the next period, and the existing tail then gets pushed back to adjoin it. But as you say, you still *eventually* lose when you get to what should have been that tail refill.

Reply:
Wait, maybe I didn't quite register everything you said, @pingerino... Why do we need to decide at kernel entry whether we're going to split the head refill? It's too late to do it in `switchSchedContext`?

I'm hoping it's better late than never to be asking these questions. :-)

Reply:
The implementaion should now delay the tail if there isn't enough unused refills in the queue for a split, the only reason the thread wouldn't still be schedulable when the kernel exits is if the head would have less the MIN_BUDGET left after charging any consumed time.

Reply:
Meta-comment: Part of the reason this PR is taking so long is there are lots of changes, and the commit messages are not explicitly stating what is happening, making it very difficult to review sufficiently. This is why what @kent-mcleod  has done is so valuable, but even in the split PR's, we are discovering new changes which we weren't aware of.

@xurtis can you explain why you thought it better to delay the tail? Does this deviate from the published algorithms? Really this should be a separate commit in itself, as it changes the scheduling behaviour. I don't recall why I chose one or the other.

@mbrcknl  we need to know at the start of the kernel if we have enough budget to do the operation we're in the kernel for, otherwise we can't do the operation and we need to simulate a timer tick. 

Example: If it's an enqueue into an IPC thread, we have to have more than MIN_BUDGET when we wake the thread up on the next kernel entry. If we check for the split at the end, we may end up in a case where we have to undo what we did, effectively looping in the kernel as we did an operation and now have to simulate the timer tick at the end.